[Deng Huaiyu] iP

NoraYUuu opened the pull request:
# Duke Update1

As you might be wondering *why you need to read this*:

> Add description for Pull Request for easy reference. ([source](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html))

That’s it.

## **functions changed**
* CRUD
1. `add`
1. `delete`
1. `list`
1. `complete`
* Error message for users
1. wrong input
1. complete absent task
1. delete absent task
* local saving for task list

```Java
public void loadFile() {
    storedList = new File(FILE_PATH);
    try {
        if (!storedList.exists()) {
            if (!storedList.getParentFile().exists()) {
                storedList.getParentFile().mkdirs();
            }
            storedList.createNewFile();
            System.out.println("Local file created.");
        }
        TaskList.loadList(storedList);
    } catch (IOException e) {
        System.out.println("error when loading file: "+ e);
    }
}
```

## **demo of functions** :star_struck:
* Task List
- [x] [T][ ] read book (a todo item)
- [x] [X] this is a complete item
- [x] [ ]this is an incomplete item

**Thanks for reading!**

Commit:
Add Gradle support

NoraYUuu committed:
level-1: greet, echo, exit added

NoraYUuu committed:
Level-2: Add tasks and function to print task list enabled.

NoraYUuu committed:
Level-3: mark as done.

NoraYUuu committed:
Level-4: todo, event, deadline have been added.

NoraYUuu committed:
A-TextUiTesting: add automated testing to test on add event/deadline/todo, done, list feature.

NoraYUuu committed:
Level-5: add exception handling for todo/event/deadline (without description or time) and done (with invalid index)

NoraYUuu committed:
Level-6: add delete feature and handle respected exceptions (invalid index, invalid format)

NoraYUuu committed:
A-enum: add enum for keywords (identify bye, list, done, delete)

NoraYUuu committed:
Add javadoc for public methods

NoraYUuu committed:
Level-8: allow tasks to be saved to the local hard disk

NoraYUuu committed:
Revert "Level-8: allow tasks to be saved to the local hard disk"

This reverts commit 7a3b5efcd2c771d0453f3be8e9ed52508f753b56.

NoraYUuu committed:
Revert "Revert "Level-8: allow tasks to be saved to the local hard disk""

This reverts commit bbbb2567f117b0369137b99897caf390d9d0ab27.

NoraYUuu committed:
Level-8: allows Duke to understand date and time and print out tasks on a specific date

NoraYUuu committed:
Merge branch 'branch-Level-8'

NoraYUuu committed:
debug Event: component time, array out of index bug fixed

NoraYUuu committed:
Rename certain class/methods to proper format

NoraYUuu committed:
Adds OOP into different classes by dividing classes into more specific groups.

NoraYUuu committed:
Adds classes to different packages.

NoraYUuu committed:
Adds test for TaskList(addTask method) and doneCommand(execute method) classes.

NoraYUuu committed:
Debug address of duke.txt

NoraYUuu committed:
Add comments for some public methods/classes.

NoraYUuu committed:
Add search function for Duke.

NoraYUuu committed:
Javadoc: add in more details

NoraYUuu committed:
Modifies code indentions

NoraYUuu committed:
Modify delete function

NoraYUuu committed:
Merge branch 'branch-A-CodingStandard'

NoraYUuu committed:
Merge branch 'branch-Level-9'

jianh0ng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,50 @@
+            this.k = k;
+        }
+
+        public String getK() {
```
Perhaps the naming could be more descriptive than K? I think it would increase readability

NoraYUuu replied:
Would change that!

jianh0ng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,76 @@
+            int c = t.getStatusIcon().equals("X") ? 1 : 0;
+            String now = t.getType() + " | " + c + " | " + t.getDescription();
+            if (!t.getType().equals("T")) {
+                now += " | "+t.getTime();
```
Minor issue with the missing spaces.

jianh0ng commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,395 @@
+    public void addTask(String t) throws DukeException{
+        String[] ss = t.split(" ");
+        switch (ss[0]) {
+            case("todo"):
```
Not sure about the indentation for switch case. I believe case should be 4 spaces to the left.

NoraYUuu replied:
Thanks for the comment! From coding standard, the case clauses should not be indented.

jianh0ng reviewed (commented):
Naming for some variables could be clearer, like "ind" and "getK", but otherwise looks good. :)

NoraYUuu committed:
Merge branch 'add-gradle-support' of ssh://github.com/NoraYUuu/ip into branch-A-Gradle

NoraYUuu committed:
Change unit test to allow gradle to run it and allow gradle to run the project.

NoraYUuu committed:
Check coding standard for existing code.

NoraYUuu committed:
Add GUI for Duke

NoraYUuu committed:
Add more functions to the GUI.

NoraYUuu committed:
Add use of vararg when processing input from user.

Li-Lehao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+        storage.loadFile();
+    }
+
+    public void processInput(String input) {
+        int result = parser.parse(input);
+        Command executeNext;
+        switch (result) {
+        case 0:
+            ExitCommand exitCommand = new ExitCommand();
+            exitCommand.execute();
+            break;
+        case 1:
+            executeNext = new ListCommand(myTasks);
+            executeNext.execute();
+            break;
+        case 2:
+            executeNext = new DoneCommand(myTasks, input);
+            executeNext.execute();
+            break;
+        case 3:
+            executeNext = new DeleteCommand(myTasks, input);
+            executeNext.execute();
+            break;
+        case 4:
+            executeNext = new GetDayCommand(input);
+            executeNext.execute();
+            break;
+        case 6:
+            executeNext = new SearchCommand(myTasks, input);
+            executeNext.execute();
+            break;
+        case 5:
+            executeNext = new AddCommand(myTasks, input);
+            executeNext.execute();
+            break;
+        default:
+            break;
+        }
+    }
```
The `processInput` function is very clear and elegant.

Li-Lehao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,50 @@
+        SEARCH("find"),
+        DONE("done");
+
+        private final String k;
```
Since `String k` is an attribute of `KeyWord`, maybe we need to make its name more precise?

Li-Lehao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,50 @@
+     * @param next next line to be assessed.
+     * @return an int indicating next step to be taken.
+     */
+    public int parse(String next) {
+        if (next.equals(KeyWord.END.getK())) {
+            return 0;
+        } else if (next.equals(KeyWord.LIST.getK())) {
+            return 1;
+        } else if (next.length() > 3 && next.substring(0, 4).equals(KeyWord.DONE.getK())) {
+            return 2;
+        } else if (next.length() > 5 && next.substring(0, 6).equals(KeyWord.DELETE.getK())) {
+            return 3;
+        } else if (next.length() > 2 && next.substring(0, 3).equals(KeyWord.GET.getK())) {
+            return 4;
+        } else if (next.length() > 3 && next.substring(0, 4).equals(KeyWord.SEARCH.getK())) {
+            return 6;
+        } else {
+            return 5;
+        }
+    }
+}
```
I really like this clean and elegant `parse` function. The implementation of `parser` is always tricky but you really did a good job. Maybe we can consider returning an `enum` type to enhance the readability of the code?

Li-Lehao commented on src/main/java/duke/command/SearchCommand.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class SearchCommand extends Command {
+    private TaskList taskList;
+    private String s;
```
Maybe we need to make its name more meaningful?

NoraYUuu committed:
Add assertion for some methods.

Add assertions for TaskList to check the number of tasks in the TaskList
is non-negative. Add assertion to parser class to ensure the string to
parse is not empty. Duke class also has an assertion to check the
parsing result to be positive integer.

Assertions check the code fulfils the expectations.

NoraYUuu committed:
Improve code quality following coding standard.

Suggestions from PR review is considered, and variable name as well as
return type of parse method are changed to enhance readability.

Arrowhead shape code is changed to guard clauses according to coding
quality instruction in week5 notes.

NoraYUuu committed:
Add stream operation.

When processing tasklist to be printed on the interface, stream is used
to map each task to the appropriate output format.

NoraYUuu committed:
Merge pull request #1 from NoraYUuu/branch-A-Assertions

Add assertion for some methods.

NoraYUuu committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

NoraYUuu committed:
Merge pull request #2 from NoraYUuu/branch-A-CodeQuality

Branch A CodeQuality

NoraYUuu committed:
Add CI to Duke.

NoraYUuu committed:
Add reminder function.

Tasks to be done by today would be printed when user open the application
 and when user uses "get" function to get tasks on a day.

"get today" instruction is added to give deadlines today quickly.

NoraYUuu committed:
Add warning box to Gui.

When a warning message is shown by duke, the format would be different
from an ordinary message.

The title of the chatbox is also changed to "TaskBot".

NoraYUuu committed:
Set theme jekyll-theme-cayman

NoraYUuu committed:
Update UserGuide.

NoraYUuu committed:
Merge branch 'master' of https://github.com/NoraYUuu/ip

NoraYUuu committed:
Update ReadME.

NoraYUuu committed:
Update testcases for Junit test.

NoraYUuu committed:
Update README.md
